### ISSUE TITLE
Add more input types and patterns to the form generator

### ISSUE THREAD

**Issue by akroii** [contao/contao#2375]
## Input types
Currently theres a lot of input-types (`type="text"`) in Contao 4 with different validation:
```
<input type="number" name="Numerische_Zeichen" id="ctrl_1" class="text" value="" step="any">
<input type="text" name="Alphabetische_Zeichen" id="ctrl_3" class="text" value="">
<input type="text" name="Erweiterte_Alphabetische_Zeichen" id="ctrl_4" class="text" value="">
<input type="text" name="Datum" id="ctrl_5" class="text" value="">
<input type="text" name="Uhrzeit" id="ctrl_6" class="text" value="">
<input type="text" name="Datum_Uhrzeit" id="ctrl_7" class="text" value="">
<input type="tel" name="Telefonnummer" id="ctrl_8" class="text" value="">
<input type="email" name="E-Mail" id="ctrl_9" class="text" value="">
<input type="url" name="URL" id="ctrl_10" class="text" value="">
```

According to the ([mozilla mdn](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input) and [w3c consortium](https://www.w3.org/wiki/HTML/Elements/input/time)) theres a few more input types available:
```
<input type="color" name="color">
<input type="datetime-local" name="datetime-local">
<input type="month" name="month">
<input type="search" name="search">
<input type="tel" name="tel">
<input type="time" name="time">
<input type="week" name="week">
```
The result is, that we can use client-side date-picker like:
<img width="621" alt="Bildschirmfoto 2020-09-29 um 11 28 28" src="https://user-images.githubusercontent.com/8830861/94540598-ed789100-0246-11eb-99ad-bd7d08c2830f.png">

Or for input type="color":
<img width="272" alt="Bildschirmfoto 2020-09-29 um 11 29 19" src="https://user-images.githubusercontent.com/8830861/94540727-10a34080-0247-11eb-9863-55e522f7a338.png">

... and placeholders like `tt.mm.yyyy` from the browser to help people to insert theirs data:
<img width="228" alt="Bildschirmfoto 2020-09-29 um 11 32 57" src="https://user-images.githubusercontent.com/8830861/94541173-8d361f00-0247-11eb-9b3b-037b985df956.png">

## Pattern
For some elements like `<input type="text">` and `<input type="date">` theres a [pattern](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#htmlattrdefpattern) attribute available to have a client-side validation.

For example:
```
<input type="text" name="name" id="name" required pattern="[A-Za-z]+">
<input type="text" pattern="([01]?[0-9]|2[0-3]):[0-5][0-9]" id="24h"/>
<input type="date" name="form_fields" pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}">
```
Maybe we can add those input-types and attributes? 😊

**There are some old issues (Contao 3.5) I found:**
[#5918](https://github.com/contao/core/issues/5918)
[#5238](https://github.com/contao/core/issues/5238)

So the question is: are attributes like date, month, week etc. still not compatible with date validations in Contao 4? [(Source)](https://github.com/contao/core/commit/9c2978bed97a5c597526a23f484d053c8cd2dd64) ?

**Comment by aschempp** [comment]
@fritzmg has already provided a pullrequest for lots of this in https://github.com/contao/contao/pull/2297

**Comment by fritzmg** [comment]
My PR only adds new server side validation options - but not new HTML5 input types. I do agree that the form generator should support all available HTML5 input types in one way or another.

**Comment by aschempp** [comment]
If I enter a custom regex field, shouldn't it be added to the widget HTML as well?

**Comment by fritzmg** [comment]
> If I enter a custom regex field, shouldn't it be added to the widget HTML as well?

I guess it should, yes. I wasn't even aware that that was a thing. However, that means that we must pre-define the limiter ourselves, since in the HTML5 attribute the regex needs to be defined without the limiter. And that also means we cannot support any modifiers?

**Comment by akroii** [comment]
There is a new extension available, maybe you can use it?
[https://github.com/e-spin/form-regex-bundle](https://github.com/e-spin/form-regex-bundle)

**Comment by aschempp** [comment]
@fritzmg hasn't this been added in 4.11?

**Comment by fritzmg** [comment]
No, there are still input types missing, as well as the pattern attribute for example.

**Comment by fkaminski** [comment]
Would be really nice to have. Especially regarding the pattern attributes. On the one hand to ensure that the validation (server-side / client-side) is identical in each case and on the other hand you then have the possibilities via CSS to style the possible client-side mismatches.

**Comment by zonky2** [comment]
I once wrote an extension to be able to use RegEx - so far I have not noticed any problems when passing the pattern to the HTML.
see https://github.com/e-spin/form-regex-bundle

**Comment by fkaminski** [comment]
@zonky2 Good to know. 
I would still like to see general support via the Core. Why? You can simply activate the client-side validation. But under certain circumstances it can happen that the client-side validation is correct and still the server-side validation takes effect afterwards .

**Comment by zonky2** [comment]
@fkaminski my extension works both on the client (optional) and on the server...

**Comment by fkaminski** [comment]
Yepp, but probably many users are not aware of it (to need an extension).

**Comment by akroii** [comment]
Yet another extension: [https://github.com/inspiredminds/contao-extended-form-fields](https://github.com/inspiredminds/contao-extended-form-fields)

**Comment by fritzmg** [comment]
https://github.com/contao/contao/pull/4936 would implement the `date` input type for Contao 5.1.
